src/tokens/reader_stream.rs

```diff
@@ -0,0 +1,560 @@
+// Copyright Amazon.com, Inc. or its affiliates.
+
+use super::{
+    AnnotationsThunk, ContainerType, Content, FieldNameThunk, Instruction, ScalarThunk, ScalarType,
+    ScalarValue, Token, TokenStream,
+};
+use crate::result::illegal_operation;
+use crate::thunk::Thunk;
+use crate::{IonReader, IonResult, IonType, StreamItem, Symbol};
+use std::cell::RefCell;
+use std::marker::PhantomData;
+use std::rc::Rc;
+
+const INVALID_TOKEN_ERR_TEXT: &str = "TokenStream is past this token";
+const INVALID_TOKEN_PANIC_TEXT: &str = "Invalid underlying reader state";
+
+// TODO make this more generic with respect to other readers--the problem is Item/Symbol
+// TODO this has to abstract over potentially system reader to implement macros
+
+/// Internal tracking state of the returned token.  This is necessary because the lifetime of the
+/// token returned from the stream is bound to the stream itself.
+///
+/// This is essentially strongly typed [`Option`] with convenience to operate within the
+/// context of a [`ReaderTokenStream`]
+enum UnderlyingReader<R>
```

Comment:
🗺️ The lifetime of `next_token` is not bound to any static lifetime we can keep track of, this is because if we want to adapt `IonReader`, `next`/`read_XXX` are not associated in any lexical context that it may be valid for as one lifetime that we can track for our `Token`.  This lets us make sure the `Reader` we "gave" to a `Token` can be invalidated when the `Reader` "moves" one while the caller may have held on to that `Token`.